libraries/tarball.rb

```diff
@@ -0,0 +1,141 @@
+# coding: UTF-8
+#
+# Cookbook Name:: cerner_splunk
+# File Name:: tarball.rb
+#
+# Methods for working with data from a tarball.
+
+require 'rubygems/package'
+require 'zlib'
+require 'fileutils'
+
+module CernerSplunk
+  # Utility class to encapsulate operations on tar.gz files
+  # Mad Props to http://stackoverflow.com/a/19139114/504685
+  class TarBall
+    def initialize(file_name, options = {})
+      @data = []
+      @clean_proc = Remover.new(@data)
+      ObjectSpace.define_finalizer(self, @clean_proc)
+      @file_name = @data[0] = file_name
+      @reader = @data[1] =  Gem::Package::TarReader.new(Zlib::GzipReader.open @file_name)
+      @prefix = options[:prefix]
+      @options = options
+    end
+
+    attr_reader :prefix
+
+    def get_file(name)
+      data = []
+      target_path = @prefix.nil? ? name : "#{@prefix}/#{name}"
+      iterate file: proc { |path, entry| data << entry.read if path == target_path }, other: proc {}
+      fail "Multiple entries for #{target_path} found in #{@file_name}" if data.size > 1
+      data.first
+    end
+
+    def extract(extract_dir)
+      file_handler = proc do |path, entry|
+        target = "#{extract_dir}/#{path}"
+        make_dirs extract_dir, path
+        File.open(target, 'wb') { |f| f.print entry.read }
+        FileUtils.chmod entry.header.mode, target
+        FileUtils.chown @options[:user], @options[:group], target
+      end
+
+      dir_handler = proc do |path, entry|
+        target = "#{extract_dir}/#{path}"
+        make_dirs extract_dir, path, true
+        FileUtils.chmod entry.header.mode, target
+      end
+
+      symlink_handler = proc do |path, entry, other = {}|
+        target = "#{extract_dir}/#{path}"
+        make_dirs extract_dir, path
+        File.unlink target if File.exist? target
+        File.symlink other[:linkname], target
+        FileUtils.chmod entry.header.mode, target
+        FileUtils.chown @options[:user], @options[:group], target
+      end
+
+      iterate file: file_handler, directory: dir_handler, symlink: symlink_handler
+    end
+
+    def count(callbacks = {}, &other)
+      callbacks = { other: other } if other && callbacks.empty?
+      i = 0
+      hsh = callbacks.each_with_object({}) { |(k, v), h| h[k] = proc { |*args| i += 1 if v.call(args) } }
```

Comment:
some doc here would help me out. I'm not sure what's going on with the callbacks and how it relates to this count that actually gets returned.